### PR TITLE
Refresh mitie sklearn and add tests

### LINKED ISSUE #137: refresh `mitie_sklearn` backend
The `mitie_sklearn` backend hasn't ever been fully supported but it's relevant to reintroduce it seeing how many people are suffering from long training times. It's basically the same as the MITIE backend, but replaces the intent classifier with the `SklearnIntentClassifier`. It's not up to date with the current API though and needs refreshing

### PR DESCRIPTION
**Pull request by plauto**
Fixes #137 

Might need to to perform a further update once https://github.com/golastmile/rasa_nlu/pull/151 is accepted in order to add support for the synonyms within the interpreter

### CODE REVIEW

**Comment by amn41** [comment]
Cool. Could you please also add a test for running the trained model?

File: _pytest/test_train.py [hunk 1]
```diff
@@ -28,6 +28,18 @@ def test_train_mitie():
     run_train(_config)
 
 
+def test_train_mitie_sklearn():
+    # basic conf
+    _config = {
+        'write': os.path.join(os.getcwd(), "rasa_nlu_logs.json"),
```
Comment (tmbo) [inline _pytest/test_train.py]: should use tempfile

File: config_mitie.json [hunk 2]
```diff
@@ -1,5 +1,5 @@
 {
-  "backend": "mitie",
+  "backend": "mitie_sklearn",
```
Comment (tmbo) [inline config_mitie.json]: is this intended?

File: src/interpreters/mitie_sklearn_interpreter.py [hunk 3]
```diff
@@ -1,12 +1,17 @@
...
 from rasa_nlu import Interpreter
+from rasa_nlu.featurizers.mitie_featurizer import MITIEFeaturizer
 from rasa_nlu.tokenizers.mitie_tokenizer import MITIETokenizer
 
 
 class MITIESklearnInterpreter(Interpreter):
-    def __init__(self, metadata):
-        self.extractor = named_entity_extractor(metadata["entity_extractor"])  # ,metadata["feature_extractor"])
-        self.classifier = text_categorizer(metadata["intent_classifier"])  # ,metadata["feature_extractor"])
+    def __init__(self, intent_classifier=None, entity_extractor=None, feature_extractor=None, **kwargs):
```
Comment (tmbo) [inline src/interpreters/mitie_sklearn_interpreter.py]: naming (I know its the same in the `MITIEInterpreter`, but lets try to improve the code base 😅): all of these three names should indicate that they are strings representing a path not the actual classifier / extractor.

File: src/interpreters/mitie_sklearn_interpreter.py [hunk 4]
```diff
@@ -18,12 +23,23 @@ def get_entities(self, tokens):
         return d
 
     def get_intent(self, tokens):
-        label, _ = self.classifier(tokens)  # don't use the score
-        return label
+        """Returns the most likely intent and its probability for the input text.
+
+        :param text: text to classify
```
Comment (tmbo) [inline src/interpreters/mitie_sklearn_interpreter.py]: there is no param `text`

File: src/interpreters/mitie_sklearn_interpreter.py [hunk 5]
```diff
@@ -18,12 +23,23 @@ def get_entities(self, tokens):
...
+        :return: tuple of most likely intent name and its probability"""
+        if self.classifier:
+            X = self.featurizer.create_bow_vecs(tokens)
```
Comment (tmbo) [inline src/interpreters/mitie_sklearn_interpreter.py]: please add a comment why tokens are passed to this function although the function expects sentences.

File: src/trainers/mitie_sklearn_trainer.py [hunk 6]
```diff
@@ -1,24 +1,37 @@
...
+        self.ensure_language_support(language_name)
 
-    def train(self, data):
+    def train(self, data, test_split_size=0.1):
         self.training_data = data
-        self.intent_classifier = self.train_intent_classifier(data.intent_examples)
-        self.entity_extractor = self.train_entity_extractor(data.entity_examples)
+        self.train_intent_classifier(data.intent_examples, test_split_size)
+
+        num_entity_examples = len([e for e in data.entity_examples if len(e["entities"]) > 0])
```
Comment (tmbo) [inline src/trainers/mitie_sklearn_trainer.py]: please add this logic as a method of `TrainingData`

File: src/trainers/mitie_sklearn_trainer.py [hunk 7]
```diff
@@ -1,24 +1,37 @@
...
         self.training_data = None
-        self.fe_file = config.mitie_file
-        self.feature_extractor = total_word_feature_extractor(self.fe_file)
+        self.fe_file = fe_file
+        self.featurizer = MITIEFeaturizer(self.fe_file)
+        self.max_num_threads = max_num_threads
+        self.ensure_language_support(language_name)
 
-    def train(self, data):
+    def train(self, data, test_split_size=0.1):
```
Comment (tmbo) [inline src/trainers/mitie_sklearn_trainer.py]: I think its time to move this function to the `Trainer` and remove the duplication in every one of the three trainers, what do you think?

File: src/trainers/mitie_sklearn_trainer.py [hunk 8]
```diff
@@ -27,46 +40,72 @@ def start_and_end(self, text_tokens, entity_tokens):
         start, end = locs[0], locs[0] + len(entity_tokens)
         return start, end
 
+    @classmethod
+    def find_entity(cls, ent, text):
```
Comment (tmbo) [inline src/trainers/mitie_sklearn_trainer.py]: This is a duplication of the code in `MITIETrainer`

File: src/trainers/mitie_sklearn_trainer.py [hunk 9]
```diff
@@ -27,46 +40,72 @@ def start_and_end(self, text_tokens, entity_tokens):
...
+            message = u"invalid entity {0} in example {1}:".format(ent, text) + \
+                u" entities must span whole tokens"
+            raise ValueError(message)
+        start = offsets.index(ent["start"])
+        _slice = text[ent["start"]:ent["end"]]
+        val_tokens = tokenize(_slice)
+        end = start + len(val_tokens)
+        return start, end
+
     def train_entity_extractor(self, entity_examples):
```
Comment (tmbo) [inline src/trainers/mitie_sklearn_trainer.py]: This is a duplication of the code in `MITIETrainer`

File: src/trainers/mitie_sklearn_trainer.py [hunk 10]
```diff
@@ -27,46 +40,72 @@ def start_and_end(self, text_tokens, entity_tokens):
...
 
         ner = trainer.train()
         return ner
 
-    def train_intent_classifier(self, intent_examples):
-        trainer = sklearn_text_categorizer_trainer(self.fe_file)
-        for example in intent_examples:
-            tokens = tokenize(example["text"])
-            trainer.add_labeled_text(tokens, example["intent"])
+    def train_intent_classifier(self, intent_examples, test_split_size=0.1):
```
Comment (tmbo) [inline src/trainers/mitie_sklearn_trainer.py]: This is a duplication of the code in `SpacySklearnTrainer`

File: src/trainers/mitie_sklearn_trainer.py (hunk 7 above)
Comment (amn41) [inline src/trainers/mitie_sklearn_trainer.py]: yep that's cleaner. We should add abstract methods for the `train_intent_classifier` & `train_entity_extractor`  to the base class as well.

File: config_mitie.json (hunk 2 above)
Comment (plauto) [inline config_mitie.json]: I think so. Technically this backend is meant to be faster, so we should consier defaulting it. 
@amn41 what do you think?

File: src/interpreters/mitie_sklearn_interpreter.py (hunk 3 above)
Comment (plauto) [inline src/interpreters/mitie_sklearn_interpreter.py]: we can append _file at the end.

File: src/interpreters/mitie_sklearn_interpreter.py (hunk 4 above)
Comment (plauto) [inline src/interpreters/mitie_sklearn_interpreter.py]: yeah now it's token, fixed. Thanks! :)

File: src/interpreters/mitie_sklearn_interpreter.py (hunk 5 above)
Comment (plauto) [inline src/interpreters/mitie_sklearn_interpreter.py]: tokens aren't passed. The function expects sentences and we actually pass those! You will see it fixed in the next commit

File: src/trainers/mitie_sklearn_trainer.py (hunk 7 above)
Comment (plauto) [inline src/trainers/mitie_sklearn_trainer.py]: no problem!

File: config_mitie.json (hunk 2 above)
Comment (amn41) [inline config_mitie.json]: ah I see. Let's make another `config_mitie_sklearn.json` . If people want to start out by installing a single dependency - that should work. We should probably throw some warnings if training is started with the MITIE backend and > 5 intents and > 50 intent examples.

**Comment by amn41** [comment]
Tom re: your comment. I agree it would be cleaner to split this up & recombine. But since there's only two cases (plain mitie and mitie_sklearn) it feels like overengineering. Maybe there's a simpler (perhaps less idiomatic) way to reduce the code duplication?

**Comment by plauto** [comment]
how about then two utils files? Like one sklearn_trainer_utils and mitie_trainer_utils each one of them with and entity and intent classifier

**Comment by tmbo** [comment]
Yes that sounds good.

File: src/trainers/mitie_sklearn_trainer.py [hunk 11]
```diff
@@ -28,45 +27,46 @@ def start_and_end(self, text_tokens, entity_tokens):
...
-                sample.add_entity(xrange(start, end), ent["entity"])
-            trainer.add(sample)
+        self.entity_extractor = mitie_trainer_utils.train_entity_extractor(entity_examples,
+                                                                           self.fe_file,
+                                                                           self.max_num_threads,)
 
-        ner = trainer.train()
-        return ner
+    def train_intent_classifier(self, intent_examples, test_split_size=0.1):
+        self.intent_classifier = SklearnIntentClassifier(max_num_threads=self.max_num_threads)
```
Comment (tmbo) [inline src/trainers/mitie_sklearn_trainer.py]: Is there a reason not to use `sklearn_trainer_utils.train_intent_classifier` here?

**Pull request merged**